processing/src/main/java/io/druid/query/ChainedExecutionQueryRunner.java

```diff
@@ -70,87 +74,115 @@ public ChainedExecutionQueryRunner(
-                                    }
-
-                                    List<T> retVal = result.toList();
-                                    if (retVal == null) {
-                                      throw new ISE("Got a null list of results! WTF?!");
-                                    }
-
-                                    return retVal;
-                                  }
-                                  catch (QueryInterruptedException e) {
-                                    throw Throwables.propagate(e);
-                                  }
-                                  catch (Exception e) {
-                                    log.error(e, "Exception with one of the sequences!");
-                                    throw Throwables.propagate(e);
-                                  }
-                                }
-                              }
-                          );
+                          return retVal;
                         }
-                    )
-                )
+                        catch (QueryInterruptedException e) {
+                          throw Throwables.propagate(e);
+                        }
+                        catch (Exception e) {
+                          log.error(e, "Exception with one of the sequences!");
+                          throw Throwables.propagate(e);
+                        }
+                      }
+                    }
+                ).map(exec::submit)
             );
 
             queryWatcher.registerQuery(query, futures);
 
             try {
+              final DateTime deadline;
+              if (QueryContexts.hasTimeout(query)) {
+                deadline = DateTimes.nowUtc().plusMillis((int) QueryContexts.getTimeout(query));
+              } else {
+                deadline = DateTimes.utc(JodaUtils.MAX_INSTANT);
+              }
+              ForkJoinPool.managedBlock(new ForkJoinPool.ManagedBlocker()
+              {
+                @Override
+                public boolean block() throws InterruptedException
+                {
+                  try {
+                    futures.get(JodaUtils.timeoutForDeadline(deadline), TimeUnit.MILLISECONDS);
+                  }
+                  catch (ExecutionException | TimeoutException e) {
+                    // Will get caught later
+                  }
+                  return true;
+                }
+
+                @Override
+                public boolean isReleasable()
+                {
+                  return futures.isDone() || deadline.isBefore(DateTimes.nowUtc());
+                }
+              });
               return new MergeIterable<>(
                   ordering.nullsFirst(),
-                  QueryContexts.hasTimeout(query) ?
-                      futures.get(QueryContexts.getTimeout(query), TimeUnit.MILLISECONDS) :
-                      futures.get()
+                  futures.get(JodaUtils.timeoutForDeadline(deadline), TimeUnit.MILLISECONDS)
```

Comment:
The timeout here has no point. Please remove it and add a comment why we call `futures.get()` without timeout.